caddyhttp/proxy/reverseproxy.go

```diff
@@ -260,8 +260,23 @@ func (rp *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request,
 		}
 		defer backendConn.Close()
 
-		go pooledIoCopy(backendConn, conn) // write tcp stream to backend
-		pooledIoCopy(conn, backendConn)    // read tcp stream from backend
+		// Proxy backend -> frontend.
+		go pooledIoCopy(conn, backendConn)
+
+		// Proxy frontend -> backend.
+		//
+		// NOTE: Hijack() sometimes returns buffered up bytes in brw which
+		// would be lost if we didn't read them out manually below.
+		if brw != nil {
+			if n := brw.Reader.Buffered(); n > 0 {
+				rbuf, err := brw.Reader.Peek(n)
+				if err != nil {
+					return err
+				}
+				backendConn.Write(rbuf)
```

Comment:
The `brw` is a alternative, low-level interface to the `conn`. If we `Read()` more from it than we have buffered up it will delegate the `Read()` to the `conn`. We don't want that, because we already do a regular `io.Copy` from the `conn` right after. What we want instead is `brw`'s internal `buf` and `Peek()` is the most straightforward choice for that (as it guarantees to only return buffered bytes). After we read those bytes out the `brw` can be garbage collected since we actually don't need it at all (apart from circumventing this design bug in Go's stdlib).